Add theme showcase banners.

mendezcode opened the pull request:
This PR adds the code to implement two different banners to be diplayed (randomly) on the theme showcase, on page load. The banners are responsive, and smoothly adjust the opacity of the image based on the width.

![screen shot 2018-01-30 at 6 38 51 pm](https://user-images.githubusercontent.com/5881557/35595981-d78c7598-05ee-11e8-9050-3a2986778fc2.png)

![screen shot 2018-01-30 at 6 39 12 pm](https://user-images.githubusercontent.com/5881557/35595984-dabe6370-05ee-11e8-970c-a223b2fedb97.png)

mapk commented:
@mendezcode Wow, this is great! 

Looks like you went with changing the opacity on the image. I didn't notice a max-wdth on the line-length like we talked. Also the PhotoBlog ad is missing the image swap for mobile screens. It was supposed to show a different image on smaller size screens, but that's okay. Let's get this live to see if we're in the right ballpark.

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -215,6 +266,7 @@ class ThemeShowcase extends React.Component {
 				) }
 				<div className="themes__content">
 					<QueryThemeFilters />
+					<Banner />
```
We don't really need the `Banner` const -- let's just put `<ThemesBanner { ...bannerProps } />` right here.

mendezcode replied:
> Also the PhotoBlog ad is missing the image swap for mobile screens. It was supposed to show a different image on smaller size screens, but that's okay. Let's get this live to see if we're in the right ballpark.

@mapk I was not aware of another image to be used for mobile screens though. This in particular, changes the whole implementation since there's another variable we need to take into account. I think we need to discuss this in more detail, since we would need to specify several images as opposed to a single one.

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -5,31 +5,56 @@
+		title: PropTypes.oneOfType( [ PropTypes.string, PropTypes.object, PropTypes.array ] )
+			.isRequired,
+		description: PropTypes.oneOfType( [ PropTypes.string, PropTypes.object, PropTypes.array ] )
+			.isRequired,
```
That seems a bit too permissive -- in what cases would `title` or `description` be an array or object?

mendezcode replied:
Good question! There are 3 things we can provide:

1) A String: `description={ "Hello World" }` 
2) An Object: `description={ <span>Hello <strong>World</strong></span> }`
3) An Array: `description={ translate( ... ) }`

The first is just a plain string, without translation. The second is a bare JSX node, which seems to be represented as an object internally. The third is the output of `translate()`.

I know it's too much, maybe we should just allow translated strings there? That seems more reallistic to me since the banner text would need to be translated anyway...

ockham replied:
Hmm, why is the result of `translate()` an array? That comes as a surprise to me...

Anyway, maybe `PropTypes.element` is the best fit here, since it allows any kind of React element; or `PropTypes.node`, which is even more permissive (allowing pretty much anything), but less verbose than `oneOfType( [ PropTypes.string, PropTypes.object, PropTypes.array ] )`.

(see [docs](https://www.npmjs.com/package/prop-types#usage) for available types)

mendezcode replied:
Oh, `PropTypes.element` should do the trick, thanks for the tip! :D

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -58,6 +60,53 @@ const optionShape = PropTypes.shape( {
+		},
+	},
+] );
+
```
I think I'd prefer defining two components (let's say `PhotgraphyThemesBanner` and `SmallBizThemesBanner`) instead of two sets of props, and choosing from them :slightly_smiling_face: What do you think?

mendezcode replied:
Yeah, that works too, but in the end we need to only be able to display one, which is chosen at the time of render. If you notice, we are doing:

```
bannerProps = sample( [ spec, spec ] )
```

It's using `lodash.sample()` to get a random _banner props object_ from the array, which is then used to render the actual banner. This is done outside the Component scope, to prevent the random component being affected by re-renders. If done inside, we would se both banners being re-rendered very fast, which would look like a roulette...

What we could do in this case is give a name to the banner props, something like:

```
const photoBlogBanner = {
	...
}

const smallBusinessBanner = {
	...
}

const bannerProps = sample( [
	photoBlogBanner,
	smallBusinessBanner,
] );```

What do you think?

ockham replied:
Hmm, my preferred semantics would be `Banner = sample( [ PhotgraphyThemesBanner, SmallBizThemesBanner ] )`. Are you saying that won't work? (If so, I don't get why not -- could you explain? :slightly_smiling_face:)

mendezcode replied:
Ah, yeah that works too!

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -58,6 +60,53 @@ const optionShape = PropTypes.shape( {
+		action: () => {
+			alert( 'Small Business banner button clicked!' );
+		},
+		actionLabel: translate( 'See the theme' ),
```
This might be the same for every theme, so we might be able to drop the prop and just hard-wire it in the `ThemesBanner` component.

mendezcode replied:
I don't think we will always have themes showcasing in the banners, we may promote a bundle in the future. Not restricting it to themes only I think is good, since we may want to promote other things like a bundle of both a plan and a theme in the future.

ockham replied:
The thing is called `ThemesBanner` tho 😬 But yeah, okay.

mendezcode replied:
Haven't thought about that but you're right, haha!

My initial thought on _Themes_, as in the Theme Team, maybe we should call it PromoBanner, to be more semantic... Any names ring the bell? that's the best one I can think of!

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -58,6 +60,53 @@ const optionShape = PropTypes.shape( {
+		),
+		action: () => {
+			alert( 'Photo Blog banner button clicked!' );
+		},
```
We don't want to merge this with the `alert`, do we? 😄

mendezcode replied:
Haha No way! It's just there until I am instructed what kind of action are we going to do for the moment, if we're going to track the clicks, then we may use the `action`, or if we just want to go to a URL, then we just replace it with a plain `href`. I'll refactor this when I get the clear direction we're going with this.

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -58,6 +60,53 @@ const optionShape = PropTypes.shape( {
+		),
+		action: () => {
+			alert( 'Small Business banner button clicked!' );
+		},
```
https://github.com/Automattic/wp-calypso/pull/21983/files#r165021287

mendezcode commented:
@ockham I made your suggested changes, thanks a lot!

On a side note, using `PropTypes.element` didn't work, I went with `PropTypes.node`, which seems to be exactly what we need in this situation.

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -58,6 +60,64 @@ const optionShape = PropTypes.shape( {
+const PhotoBlogBanner = () => {
+	const onClick = () => {
+		alert( 'Photo Blog banner button clicked!' );
+	};
```
If we're adding an action, this shouldn't be a functional component, but rather `extend PureComponent`. However, you might just want to drop the dummy actions for now since we don't want to merge this PR with them anyway (and change to a `PureComponent` later, if/when needed).

mendezcode commented:
Hey @ockham, I refactored and made the buttons go to the theme pages, and added tracking of the banner clicks as well. I also changed the banners to be _Pure Components_, let me know what you think! :)

Note: There's a change coming up, the entire banner should be made clickable as per @mapk's suggestion.

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -58,6 +60,98 @@ const optionShape = PropTypes.shape( {
+			getCurrentTheme: curry( getActiveTheme )( state ),
+		} ) )( component )
+	)
+);
```
Wow, quite complex 😅 Instead of doing the same thing for each individual banner, I'd much rather suggest doing it once to the `ThemesBanner` component that both are based on 😉 (right in `themes-banner/index.jsx`).

ockham replied:
(I'd suggest making `themeId` a `ThemesBanner` prop to that end.)

ockham replied:
Also, we generally try to avoid passing (bound) functions as props from `connect()` to the wrapped component...

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -58,6 +60,98 @@ const optionShape = PropTypes.shape( {
+	const tracksData = {
+		site_id: siteId,
+		current_theme: getCurrentTheme( siteId ),
+		timestamp: new Date().toGMTString(),
```
Do we need to do this? I was kind of assuming tracks would take care of adding something like a timestamp for us automagically. I might need to verify that assumption...

ockham replied:
![image](https://user-images.githubusercontent.com/96308/35880832-dbd852b0-0b76-11e8-9020-3dea1b5ffb55.png)

Looks like it does :slightly_smiling_face:

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -58,6 +60,98 @@ const optionShape = PropTypes.shape( {
+		timestamp: new Date().toGMTString(),
+	};
+	recordTracksEvent( tracksEvent, tracksData );
+	page( url );
```
Can we avoid the `page` call and keep an `href` prop on `ThemesBanner`? (`href` and `action` aren't mutually exclusive e.g. for our `Button` component, so we can that behavior here I think.)

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -58,6 +60,98 @@ const optionShape = PropTypes.shape( {
+		} ) )( component )
+	)
+);
+
```
These are quite a few lines of code added -- I think we should move them off `theme-showcase.jsx` to a separate file.

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -58,6 +60,98 @@ const optionShape = PropTypes.shape( {
+
+	render() {
+		const onClick = () => {
+			bannerClickHandler( 'photo-blog', 'calypso_showcase_banner_photo_blog_click', this.props );
```
Instead of making the theme slug part of the event name, what about keeping the event name theme-agnostic (i.e. `calypso_showcase_banner_click`), and passing the theme slug as part of the event date?

(Also, maybe the event name should include `theme`, i.e. `calypso_theme_showcase_banner_click`.)

ockham replied:
This would also allow us to move the event tracking `onClick` handler to the underlying `ThemesBanner`, which makes more sense IMO.

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -5,31 +5,56 @@
+		const imageNode = image ? (
+			<img
+				alt=""
+				{ ...imageAttrs }
```
Since it's not totally obvious that `...imageAttrs` contains an `alt` attr, we could add a `themeName` prop to `ThemesBanner` that we use to set `alt={ translate( '%(theme)s Theme, args: { theme: themeName } ) }`.

ockham commented on client/my-sites/themes/themes-banner/README.md:
```diff
@@ -5,32 +5,47 @@ This component is used to implement a banner with image, text, and button.
+				themeId={ 'small-business' }
+				themeName={ 'Small Business' }
+				title={ translate( 'Do you run a small business?' ) }
+				description={ 'An awesome description for this theme' }
```
Maybe wrap in a `translate` call? (Not too relevant, since it's just a README example, but it might be helpful for anyone who copy-pastes this.)

mendezcode replied:
True, good catch! Will update accordingly.

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -3,34 +3,105 @@
+			theme: themeId,
+			active_theme: activeThemeId,
+		};
+		recordTracksEvent( 'calypso_showcase_banner_click', tracksData );
```
👍

ockham replied:
Oops, I missed that we were directly _calling_ the `recordTracksEvent`  action here. That won't work; instead, we have to `dispatch()` it. The preferred way to do that in a React component is to use `connect()`'s second argument, [`mapDispatchToProps`](https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options), and use the "object shorthand" there.

I'll follow up with a code sample...

ockham replied:
```diff
diff --git a/client/my-sites/themes/themes-banner/index.jsx b/client/my-sites/themes/themes-banner/index.jsx
index 072be31..bfafcdb 100644
--- a/client/my-sites/themes/themes-banner/index.jsx
+++ b/client/my-sites/themes/themes-banner/index.jsx
@@ -16,7 +16,7 @@ import PropTypes from 'prop-types';
 import Button from 'components/button';
 import safeImageUrl from 'lib/safe-image-url';
 import { getSelectedSiteId } from 'state/ui/selectors';
-import { recordTracksEvent } from 'state/analytics/actions';
+import { recordTracksEvent as recordTracksEventAction } from 'state/analytics/actions';
 import { getActiveTheme, getThemeDetailsUrl } from 'state/themes/selectors';
 
 class ThemesBanner extends PureComponent {
@@ -37,7 +37,7 @@ class ThemesBanner extends PureComponent {
        };
 
        recordEvent = () => {
-               const { siteId, themeId, activeThemeId } = this.props;
+               const { siteId, themeId, activeThemeId, recordTracksEvent } = this.props;
                const tracksData = {
                        site_id: siteId,
                        theme: themeId,
@@ -99,4 +99,6 @@ const mapStateToProps = ( state, { themeId } ) => {
        };
 };
 
-export default connect( mapStateToProps )( localize( ThemesBanner ) );
+export default connect( mapStateToProps, {
+       recordTracksEvent: recordTracksEventAction,
+} )( localize( ThemesBanner ) );
```

(untested)

ockham replied:
So this essentially just adds an object as `connect`'s second argument, whose values will be bound to `dispatch()` by `connect()`. The `dispatch()`-bound actions will then be passed on to the wrapped component as props (using the object keys as prop names).

ockham replied:
The rest of the diff essentially makes sure that the `dispatch()`-bound `recordTracksEvent` action prop name doesn't shadow the unbound one that we import from `state/analytics/actions`.

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -3,34 +3,105 @@
+		activeThemeId: PropTypes.string,
+	};
+
+	onClick = () => {
```
High-level nit: I've been calling my handlers like that 😄  until someone suggested I use a more descriptive name (to illustrate what the method actually does, not what it's used for). So maybe rename it to something like `recordEvent`?

mendezcode replied:
Interesting point, will update accordingly. Makes absolute sense!

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -3,34 +3,105 @@
+		themeUrl,
+		activeThemeId,
+	};
+};
```
Nice! Looks much cleaner like this 👍

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -3,34 +3,105 @@
+		const backgroundStyle = backgroundColor ? { backgroundColor } : {};
+		const imageNode = image ? (
+			<img
+				{ ...imageAttrs }
```
What about going with YAGNI and only using an `imageWidth` prop for now?

mendezcode replied:
Yeah, I think YAGNI makes more sense here. Good little detail! :)

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -3,34 +3,105 @@
+				<h1 className="themes-banner__title">{ title }</h1>
+				<p className="themes-banner__description">{ description }</p>
+				<Button className="themes-banner__cta" compact primary>
+					{ buttonLabel }
```
As discussed during our call, if we confine the banner to themes (as is reflected in the name we're using now -- `ThemesBanner`) and aren't going to use it for different purposes (such as upgrades/plans etc.), it makes sense to drop the `buttonLabel` prop and just hard-wire it as `translate( 'See the theme' )`.

mendezcode replied:
Yeah, that can be done.

What if I make it optional and have it default to `See the theme`?

ockham replied:
> But, enforcing YAGNI I think it's best to hardcode it for now, if we need a new label then we change accordingly.
> 
> No need to have unnecessary complexity or flexibility, things can change later and evolve as the needs arise.

Yeah, that'd be my line of thinking here. If the need arises later, we can always add the prop (maybe with a default, as you're suggesting). Better to start simple and add complexity only when needed IMO 😄

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -3,34 +3,105 @@
+					{ buttonLabel }
 				</Button>
-			</div>
+				{ imageNode }
```
I think it's fine to just inline the `{ image ? ( <img /*...*/ /> ) : null }` tertiary here (actually, `{ image && ( <img /*...*/ /> ) }` might work as well).

mendezcode replied:
Perfect, will do! I thought about it actually...

ockham commented on client/my-sites/themes/themes-banner/photo-blog.jsx:
```diff
@@ -0,0 +1,44 @@
+class PhotoBlogBanner extends PureComponent {
+	render() {
+		const { translate } = this.props;
+		return (
```
This is a total matter of taste, so feel free to ignore the suggestion:
I personally like to use functional stateless components in simple cases like this, since they express so nicely the mapping from an input (a set of props) to a result (a component). Are you using the `PureComponent` because it'll easier to add methods to it? Or any other particular reason? (Really just curious 😄)

(Functional stateless equivalent would e.g. be
```js
const PhotoBlogBanner = ( { translate } ) => (
	<ThemesBanner
		themeId="photo-blog"
		/*...*/
	/>
);
```

mendezcode replied:
Actually, this approach is so much cleaner and expressive, straight to the point, no fuss...

It's just a simple function returning the component instance. No need to require `PureComponent` or declare extra `render()` functions, etc. Very concise, I love it! :)

ockham commented on client/my-sites/themes/themes-banner/photo-blog.jsx:
```diff
@@ -0,0 +1,44 @@
+		const { translate } = this.props;
+		return (
+			<ThemesBanner
+				themeId={ 'photo-blog' }
```
For string literals in JSX, double quotes are preferred over curly-braces-and-single-quotes, i.e. `themeId="photo-blog"` (same for `themeName`).

ockham replied:
(And also for `backgroundColor`, `image`, and `imageTransform`.)

mendezcode replied:
Good to know!

mendezcode commented:
@ockham This review has been amazing! Love all of the points you brought up, thank you so much!

I have implemented everything as suggested. It all makes a ton of sense! ;-)

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -58,6 +60,8 @@ const optionShape = PropTypes.shape( {
 	action: PropTypes.func,
 } );
 
+const ShowcaseBanner = sample( [ PhotoBlogBanner, SmallBusinessBanner ] );
```
There's one more thing that occurred to me: The Theme Showcase is server-side rendered. Adding this random element will cause reconciliation errors whenever client-side rendering ends up picking the 'other' banner :slightly_frowning_face:

ockham replied:
As it goes with reconciliation errors, we can't also just skip the banner when rendering on the server -- since that will _also_ cause a renconc error 😄 

One option would be to only render the banner when logged-in, since we don't SSR there. That would fix the reconc issue, but it would also mean that logged-out users would not get to see the banner at all -- which might be undesired behavior.

ockham replied:
Current issue (seen on http://calypso.localhost:3000/themes while logged-out):
![image](https://user-images.githubusercontent.com/96308/36112850-d35d9356-1022-11e8-9bb6-6c7e8dc780bf.png)

ockham commented on client/my-sites/themes/themes-banner/small-business.jsx:
```diff
@@ -0,0 +1,36 @@
+		backgroundColor="#3d596d"
+		image="/calypso/images/themes-banner/small-business.png"
+		imageTransform="translateY(-19%) translateX(17%)"
+		imageWidth={ 410 }
```
I _think_ you can even use double quotes from numbers, i.e. `imageWidth="410"`.

folletto commented:
Would be possible to have a screenshot of the full page, and a pointer of _why_ this change? It's difficult to evaluate this change in isolation.

mapk commented:
@folletto Here's a screenshot of the banners in place.

![screen shot 2018-02-12 at 9 40 04 pm](https://user-images.githubusercontent.com/617986/36130528-1fe289e4-1022-11e8-83f8-9322edb4d808.png)
![screen shot 2018-02-12 at 9 40 33 pm](https://user-images.githubusercontent.com/617986/36130529-211b9cce-1022-11e8-8da3-c7cd97059506.png)

The _why_ for this change is on ProdPad here: https://app.prodpad.com/ideas/112/canvas

**What problem are you trying to solve?**
Customers experience difficulty finding the right theme for their site.

**The bet**
We bet by adding a banner to the showcase, featuring select themes, customers will pick newer themes instead older ones.

folletto commented:
> The why for this change is on ProdPad here

I've no access to that. Why it's not on a P2?

Regardless, got it  :)

The only feedback I can give is that it seems to me the style to be a bit different from Calypso (text shadows, color contrast with the button, uppercase, background color). It doesn't seem to fit... but maybe that's how it's meant to be?

mapk commented:
> I've no access to that. Why it's not on a P2?
Oh, dang, sorry about that. Most of what I included in the comment is what you would find there. So you got the gist of it. :)

> The only feedback I can give is that it seems to me the style to be a bit different from Calypso (text shadows, color contrast with the button, uppercase, background color). It doesn't seem to fit... but maybe that's how it's meant to be?
I really appreciate this! There are a few intentional design choices made so that it doesn't just blend in with Calypso. We want it to stand out. I tried to follow the new marketing design styles with the understanding that this is a marketing venture.

If you have some alternative ideas around this I'm totally up for revisions! Thanks for chiming in, @folletto!

folletto commented:
> Oh, dang, sorry about that. Most of what I included in the comment is what you would find there. So you got the gist of it. :)

Ok, thanks! :)

> We want it to stand out. I tried to follow the new marketing design styles with the understanding that this is a marketing venture.

Makes sense :)
